HIP/hip-000.md

```diff
@@ -0,0 +1,116 @@
+
+The EVM uses the nonce value internally, for several different operations, for example:
+    1. to check whether an account exists at an address, it checks whether the nonce property stored for this address is different than 0 (it also checks the balance and code properties for the address as well, but this is out of scope for this discussion).
+    2. when deploying new contracts via a `CREATE` operation, the address for an Ethereum contract is **deterministically** computed from the address of its creator (`sender`) and how many transactions the creator has sent (`nonce`). The `sender` and `nonce` are RLP encoded and then hashed with Keccak-256.
+
+Ethereum nodes expose a JSON-RPC endpoint that can be queried for each account’s nonce value: [eth_getTransactionCount](https://ethereum.org/en/developers/docs/apis/json-rpc/#eth_gettransactioncount)
+
+In Hedera there is `JSON-RPC Relay` (as described in [HIP-482](https://hips.hedera.com/hip/hip-482)) that provides the same `eth_getTransactionCount` JSON-RPC endpoint, but currently it always returns a value of 1 for contract accounts. The desired behavior would be to have a nonce value of `1 + n` where `n` is the number of times the contract created other contracts through `CREATE` or `CREATE2` operations.
+
+## Rationale
+
+As described in [HIP-482](https://hips.hedera.com/hip/hip-482), the `JSON-RPC Relay` works with `EthereumTransactions` that it forwards to the consensus nodes. Apart from the `EthereumTransaction` type the consensus nodes can also handle other types of transactions (e.g. `ContractCall` or `ContractCreate`) that are not coming from the `JSON-RPC Relay` but can be used by developers by other means.
+In general, when we talk about account nonces we will need to consider all of those transaction types for both types of accounts (i.e. EOAs and contracts):
+
+- Contract nonce updates for both cases of `EthereumTransactions` and `non-EthereumTransactions` (i.e. `ContractCall` or `ContractCreate`) will need to be tracked and externalized to Mirror Node
+- Tracking nonces for EOAs for `EthereumTransactions` is not changed (currently we don't have a specific protobuf to externalize the nonce but the nonce value is passed as part of the `EthereumTransaction` body, so Mirror Node can read the value from there)
+- Tracking nonces for EOAs for `non-EthereumTransactions` is not supported because there is no use case for it in Hedera
+
+## User stories
+
+1. As a user, when I make a request to `eth_getTransactionCount` endpoint I want to be able to get correct nonce values for contracts.
+  
+## Specification
+
+No changes to Services state will be needed because currently the nonces are currently being computed correctly there. We need to externalize these changes to Mirror Node so that it can keep a correct state representation for contract accounts.
```

Comment:
So services has always been tracking the contract nonce and will have an accurate existing value? So the first time a mirror node gets a contract nonce it could be some high value (2+) since that contract could've created many child contracts previously.

Reply:
> So services has always been tracking the contract nonce and will have an accurate existing value?

Yes, it looks like we increment the nonce in [ContractCreateTransitionLogic](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/contract/ContractCreateTransitionLogic.java#L209), [ContractCallTransitionLogic](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/txns/contract/ContractCallTransitionLogic.java#L148) and [AbstractEvmRecordingCreateOperation](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/contracts/operations/AbstractEvmRecordingCreateOperation.java#L122).

> So the first time a mirror node gets a contract nonce it could be some high value (2+) since that contract could've created many child contracts previously.

Yes, given the above, I think that should be the case.